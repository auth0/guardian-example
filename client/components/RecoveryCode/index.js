/* global confirm */

import React from 'react'
import Code from './code'

function confirmRegeneration(handleRegenerateRecoveryCode) {
  return function confirmAndRegenerate() {
    if (confirm('Are you sure you want to regenerate the recovery code?')) {
      handleRegenerateRecoveryCode()
    }
  }
}

function displayRegenerateDiv(handleRegenerateRecoveryCode, enabled) {
  return (<div>
    <p>
      The recovery code can be used to access your account in the event you lose
      access to your device.
    </p>
    <p>If you generate a new recovery code the old one will stop working.</p>
    <button
      className='btn btn-danger pull-right' disabled={!enabled}
      onClick={confirmRegeneration(handleRegenerateRecoveryCode)}
    >Regenerate Recovery Code</button>
  </div>)
}

function displayCode(recoveryCode) {
  return (<div>
    <p className='alert alert-info'>
      Your new recovery code is: <Code recoveryCode={recoveryCode} />
    </p>
    <p className='alert alert-warning'>
      Record this code, it will only be visible now and cannot be retrieved again.
      No previous code can be used.
    </p>
  </div>)
}

export default function RecoveryCode({ recoveryCode, enabled, handleRegenerateRecoveryCode }) {
  return recoveryCode
    ? displayCode(recoveryCode)
    : displayRegenerateDiv(handleRegenerateRecoveryCode, enabled)
}

RecoveryCode.propTypes = {
  handleRegenerateRecoveryCode: React.PropTypes.func.isRequired,
  recoveryCode: React.PropTypes.string,
}
