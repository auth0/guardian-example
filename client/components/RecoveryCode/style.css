.new-recovery-code {
  background: transparent;
  border: none;
  color: black;
  display: inline-block;
  lineHeight: 116%;
  textAlign: center;
  width: 25em;
}
