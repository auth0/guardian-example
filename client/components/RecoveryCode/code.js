import React from 'react'
import style from './style.css'

const formatCode = code => code.match(/(.{1,4})/g).join(' - ')

function handleClick(event) {
  const element = event.target
  if (element.setSelectionRange) {
    element.setSelectionRange(0, element.value.length)
  }
}

export default function Code({ recoveryCode }) {
  return (
    <input
      className={style['new-recovery-code']}
      readOnly
      onClick={handleClick}
      value={formatCode(recoveryCode)}
    />
  )
}

Code.propTypes = {
  recoveryCode: React.PropTypes.string,
}
