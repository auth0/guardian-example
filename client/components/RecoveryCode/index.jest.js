/* eslint-env jest */
import React from 'react'
import renderer from 'react-test-renderer'
import RecoveryCode from '.'

describe('RecoveryCode', () => {
  it('displays button to regenerate', () => {
    const component = renderer.create(
      <RecoveryCode
        enabled
        handleRegenerateRecoveryCode={() => {}}
      />
    )

    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('displays new code', () => {
    const component = renderer.create(
      <RecoveryCode
        enabled
        recoveryCode='ABCDEFGHIJKLMNOPQRSTUVWX'
        handleRegenerateRecoveryCode={() => {}}
      />
    )

    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
