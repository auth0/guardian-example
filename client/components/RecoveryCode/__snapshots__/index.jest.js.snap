exports[`RecoveryCode displays button to regenerate 1`] = `
<div>
  <p>
    The recovery code can be used to access your account in the event you lose access to your device.
  </p>
  <p>
    If you generate a new recovery code the old one will stop working.
  </p>
  <button
    className="btn btn-danger pull-right"
    disabled={false}
    onClick={[Function confirmAndRegenerate]}>
    Regenerate Recovery Code
  </button>
</div>
`;

exports[`RecoveryCode displays new code 1`] = `
<div>
  <p
    className="alert alert-info">
    Your new recovery code is: 
    <input
      onClick={[Function handleClick]}
      readOnly="true"
      style={
        Object {
          "background": "transparent",
          "border": "none",
          "color": "black",
          "display": "inline-block",
          "lineHeight": "116%",
          "textAlign": "center",
          "width": "25em"
        }
      }
      value="ABCD - EFGH - IJKL - MNOP - QRST - UVWX" />
  </p>
  <p
    className="alert alert-warning">
    Record this code, it will only be visible now and cannot be retrieved again. No previous code can be used.
  </p>
</div>
`;
